src/core/emsesp.cpp

```diff
@@ -846,41 +880,55 @@ std::string EMSESP::pretty_telegram(std::shared_ptr<const Telegram> telegram) {
     std::string src_name("");
     std::string dest_name("");
     std::string type_name("");
+
+    // Single loop to find all device information
+    bool src_found  = false;
+    bool dest_found = false;
+    bool type_found = false;
+
     for (const auto & emsdevice : emsdevices) {
-        // get src & dest
-        if (emsdevice->is_device_id(src)) {
-            src_name = emsdevice->device_type_name();
-        } else if (emsdevice->is_device_id(dest)) {
-            dest_name = emsdevice->device_type_name();
-        }
-        // get the type name
-        if (type_name.empty()) {
+        // get src name
+        if (!src_found && emsdevice->is_device_id(src)) {
+            src_name  = emsdevice->device_type_name();
+            src_found = true;
+        }
+
+        // get dest name
+        if (!dest_found && emsdevice->is_device_id(dest)) {
+            dest_name  = emsdevice->device_type_name();
+            dest_found = true;
+        }
+
+        // get the type name (try primary conditions first)
+        if (!type_found) {
             if ((telegram->operation == Telegram::Operation::RX_READ && emsdevice->is_device_id(dest))
                 || (telegram->operation != Telegram::Operation::RX_READ && dest == 0 && emsdevice->is_device_id(src))
                 || (telegram->operation != Telegram::Operation::RX_READ && src == EMSbus::ems_bus_id() && emsdevice->is_device_id(dest))) {
                 type_name = emsdevice->telegram_type_name(telegram);
+                if (!type_name.empty()) {
+                    type_found = true;
+                }
             }
         }
-    }
-    if (type_name.empty()) {
-        // fallback, get the type name from src
-        for (const auto & emsdevice : emsdevices) {
-            if (telegram->operation != Telegram::Operation::RX_READ && emsdevice->is_device_id(src)) {
-                type_name = emsdevice->telegram_type_name(telegram);
-                break;
-            }
+
+        // Early exit if we found everything
+        if (src_found && dest_found && type_found) {
+            break;
         }
     }
 
-    if (type_name.empty()) {
-        // 2nd fallback, get the type name from dest
+    // Fallback for type name if not found - try src or dest
+    if (!type_found && telegram->operation != Telegram::Operation::RX_READ) {
         for (const auto & emsdevice : emsdevices) {
-            if (telegram->operation != Telegram::Operation::RX_READ && emsdevice->is_device_id(dest)) {
+            if (emsdevice->is_device_id(src) || emsdevice->is_device_id(dest)) {
```

Comment:
This is not the same as before. Now we get the first entry for scr or dest, Before src have priority. 
I think the 0x2CC with different meaning/names for RC300 and boiler is a bug, But we have it in code as reported by users.

Reply:
thanks, I'll remove the `emsdevice->is_device_id(src)`